File: worker/src/protocol/http_payload.ts
export interface HTTPReqPayload {
  password: string;
  method: string;
  url: string;
  headers: [string, string][];
  body: Uint8Array;
}

export interface HTTPRespPayload {
  status: number;
  headers: [string, string][];
  body: Uint8Array;
}

function pushU16(out: number[], n: number): void {
  out.push((n >> 8) & 0xff, n & 0xff);
}

function pushU32(out: number[], n: number): void {
  out.push((n >> 24) & 0xff, (n >> 16) & 0xff, (n >> 8) & 0xff, n & 0xff);
}

function pushString(out: number[], s: string): void {
  const bytes = new TextEncoder().encode(s);
  pushU16(out, bytes.length);
  for (const b of bytes) {
    out.push(b);
  }
}

export function encodeHTTPReqPayload(p: HTTPReqPayload): Uint8Array {
  const out: number[] = [];
  pushString(out, p.password);
  pushString(out, p.method);
  pushString(out, p.url);
  pushU16(out, p.headers.length);
  for (const [k, v] of p.headers) {
    pushString(out, k);
    pushString(out, v);
  }
  pushU32(out, p.body.length);
  for (const b of p.body) {
    out.push(b);
  }
  return new Uint8Array(out);
}

export function decodeHTTPReqPayload(data: Uint8Array): HTTPReqPayload {
  const view = new DataView(data.buffer, data.byteOffset, data.byteLength);
  let off = 0;

  const readString = (): string => {
    if (off + 2 > data.length) {
      throw new Error("missing string length");
    }
    const len = view.getUint16(off, false);
    off += 2;
    if (off + len > data.length) {
      throw new Error("string exceeds payload");
    }
    const str = new TextDecoder().decode(data.slice(off, off + len));
    off += len;
    return str;
  };

  const password = readString();
  const method = readString();
  const url = readString();

  if (off + 2 > data.length) {
    throw new Error("missing header count");
  }
  const headerCount = view.getUint16(off, false);
  off += 2;

  const headers: [string, string][] = [];
  for (let i = 0; i < headerCount; i++) {
    headers.push([readString(), readString()]);
  }

  if (off + 4 > data.length) {
    throw new Error("missing body length");
  }
  const bodyLen = view.getUint32(off, false);
  off += 4;
  if (off + bodyLen > data.length) {
    throw new Error("invalid body length");
  }
  const body = data.slice(off, off + bodyLen);
  return { password, method, url, headers, body };
}

export function encodeHTTPRespPayload(p: HTTPRespPayload): Uint8Array {
  const out: number[] = [];
  pushU16(out, p.status);
  pushU16(out, p.headers.length);
  for (const [k, v] of p.headers) {
    pushString(out, k);
    pushString(out, v);
  }
  pushU32(out, p.body.length);
  for (const b of p.body) {
    out.push(b);
  }
  return new Uint8Array(out);
}

export function decodeHTTPRespPayload(data: Uint8Array): HTTPRespPayload {
  const view = new DataView(data.buffer, data.byteOffset, data.byteLength);
  if (data.length < 4) {
    throw new Error("http resp payload too short");
  }
  const status = view.getUint16(0, false);
  const headerCount = view.getUint16(2, false);
  let off = 4;

  const readString = (): string => {
    if (off + 2 > data.length) {
      throw new Error("missing string length");
    }
    const len = view.getUint16(off, false);
    off += 2;
    if (off + len > data.length) {
      throw new Error("string exceeds payload");
    }
    const str = new TextDecoder().decode(data.slice(off, off + len));
    off += len;
    return str;
  };

  const headers: [string, string][] = [];
  for (let i = 0; i < headerCount; i++) {
    headers.push([readString(), readString()]);
  }

  if (off + 4 > data.length) {
    throw new Error("missing body length");
  }
  const bodyLen = view.getUint32(off, false);
  off += 4;
  if (off + bodyLen > data.length) {
    throw new Error("invalid body length");
  }
  const body = data.slice(off, off + bodyLen);
  return { status, headers, body };
}

File: protocol/http_payload.go
package protocol

import (
	"encoding/binary"
	"errors"
	"fmt"
)

// HTTPReqPayload is the HTTP_REQ frame body.
type HTTPReqPayload struct {
	Password string
	Method   string
	URL      string
	Headers  [][2]string
	Body     []byte
}

// HTTPRespPayload is the HTTP_RESP frame body.
type HTTPRespPayload struct {
	Status  uint16
	Headers [][2]string
	Body    []byte
}

// EncodeHTTPReqPayload serializes an HTTP request for worker fetch.
func EncodeHTTPReqPayload(p HTTPReqPayload) []byte {
	headerBytes := encodeHeaderPairs(p.Headers)
	bodyLen := len(p.Body)
	buf := make([]byte, 0, 256+bodyLen+len(headerBytes))
	buf = appendString(buf, p.Password)
	buf = appendString(buf, p.Method)
	buf = appendString(buf, p.URL)
	buf = binary.BigEndian.AppendUint16(buf, uint16(len(p.Headers)))
	buf = append(buf, headerBytes...)
	buf = binary.BigEndian.AppendUint32(buf, uint32(bodyLen))
	if bodyLen > 0 {
		buf = append(buf, p.Body...)
	}
	return buf
}

// DecodeHTTPReqPayload deserializes an HTTP_REQ payload.
func DecodeHTTPReqPayload(data []byte) (HTTPReqPayload, error) {
	var p HTTPReqPayload
	off := 0
	var n int
	var err error

	if p.Password, n, err = readString(data, off); err != nil {
		return p, fmt.Errorf("decode http req password: %w", err)
	}
	off += n
	if p.Method, n, err = readString(data, off); err != nil {
		return p, fmt.Errorf("decode http req method: %w", err)
	}
	off += n
	if p.URL, n, err = readString(data, off); err != nil {
		return p, fmt.Errorf("decode http req url: %w", err)
	}
	off += n
	if off+2 > len(data) {
		return p, errors.New("decode http req: missing header count")
	}
	headerCount := int(binary.BigEndian.Uint16(data[off : off+2]))
	off += 2
	if p.Headers, n, err = readHeaderPairs(data, off, headerCount); err != nil {
		return p, fmt.Errorf("decode http req headers: %w", err)
	}
	off += n
	if off+4 > len(data) {
		return p, errors.New("decode http req: missing body length")
	}
	bodyLen := int(binary.BigEndian.Uint32(data[off : off+4]))
	off += 4
	if bodyLen < 0 || off+bodyLen > len(data) {
		return p, errors.New("decode http req: invalid body length")
	}
	if bodyLen > 0 {
		p.Body = make([]byte, bodyLen)
		copy(p.Body, data[off:off+bodyLen])
	}
	return p, nil
}

// EncodeHTTPRespPayload serializes an HTTP response from worker fetch.
func EncodeHTTPRespPayload(p HTTPRespPayload) []byte {
	headerBytes := encodeHeaderPairs(p.Headers)
	bodyLen := len(p.Body)
	buf := make([]byte, 2+2+len(headerBytes)+4+bodyLen)
	binary.BigEndian.PutUint16(buf[0:2], p.Status)
	binary.BigEndian.PutUint16(buf[2:4], uint16(len(p.Headers)))
	copy(buf[4:], headerBytes)
	off := 4 + len(headerBytes)
	binary.BigEndian.PutUint32(buf[off:off+4], uint32(bodyLen))
	off += 4
	if bodyLen > 0 {
		copy(buf[off:], p.Body)
	}
	return buf
}

// DecodeHTTPRespPayload deserializes an HTTP_RESP payload.
func DecodeHTTPRespPayload(data []byte) (HTTPRespPayload, error) {
	var p HTTPRespPayload
	if len(data) < 4 {
		return p, errors.New("decode http resp: payload too short")
	}
	p.Status = binary.BigEndian.Uint16(data[0:2])
	headerCount := int(binary.BigEndian.Uint16(data[2:4]))
	off := 4
	var n int
	var err error
	if p.Headers, n, err = readHeaderPairs(data, off, headerCount); err != nil {
		return p, fmt.Errorf("decode http resp headers: %w", err)
	}
	off += n
	if off+4 > len(data) {
		return p, errors.New("decode http resp: missing body length")
	}
	bodyLen := int(binary.BigEndian.Uint32(data[off : off+4]))
	off += 4
	if bodyLen < 0 || off+bodyLen > len(data) {
		return p, errors.New("decode http resp: invalid body length")
	}
	if bodyLen > 0 {
		p.Body = make([]byte, bodyLen)
		copy(p.Body, data[off:off+bodyLen])
	}
	return p, nil
}

func appendString(buf []byte, s string) []byte {
	b := []byte(s)
	buf = binary.BigEndian.AppendUint16(buf, uint16(len(b)))
	return append(buf, b...)
}

func readString(data []byte, off int) (string, int, error) {
	if off+2 > len(data) {
		return "", 0, errors.New("missing string length")
	}
	length := int(binary.BigEndian.Uint16(data[off : off+2]))
	off += 2
	if off+length > len(data) {
		return "", 0, errors.New("string exceeds payload")
	}
	return string(data[off : off+length]), 2 + length, nil
}

func encodeHeaderPairs(headers [][2]string) []byte {
	var buf []byte
	for _, h := range headers {
		buf = appendString(buf, h[0])
		buf = appendString(buf, h[1])
	}
	return buf
}

func readHeaderPairs(data []byte, off int, count int) ([][2]string, int, error) {
	start := off
	headers := make([][2]string, 0, count)
	for i := 0; i < count; i++ {
		key, n, err := readString(data, off)
		if err != nil {
			return nil, 0, err
		}
		off += n
		val, n, err := readString(data, off)
		if err != nil {
			return nil, 0, err
		}
		off += n
		headers = append(headers, [2]string{key, val})
	}
	return headers, off - start, nil
}
